enable lambda-lifting for the JavaScript target

zerbina opened the pull request:
## Summary

Previously, inner closure procedures were not lifted into free-standing procedures during `transf`, but rather kept as-is and mapped to nested JavaScript functions by `jsgen`. Besides requiring the JavaScript target to be special-cased in `transf` and `lambdalifting`, this also caused observable semantic issues: the destructors for captured variables were called when leaving the scope in which they were defined, not when the closure got destroyed.

In order to bring the closure semantics in line with the other targets and as a prerequisite for full ORC support, lambda-lifting is now also enabled when using the JS target -- the code generator is adjusted to support the lifted procedures. The code generator changes are also a
preparation for full closure iterator support.

As a side-effect, closures now also work as expected when using the JS target and executing non-compile-time-only procedures that have inner closure procedures at compile-time.

## Details

- remove the dispatching based on the selected backend from `transf` and `lambdalifting`
- remove all code related to the previous closure support from `jsgen`
- implement support for lifted procedures in `jsgen`

The hidden environment parameter of routines using the `.closure` calling convention is mapped to the `this` parameter at the JavaScript level.

When a NimSkull closure object is created, a new function instance is created via `bind` and the provided environment bound to the `this` parameter. In addition, the both the function and environment are also assigned to the respective `prc` and `env` properties on the function object -- this allows for later retrieval.

Compared to using an object, this approach has the benefit that closures can still be passed to an imported procedure that expects a JavaScript function object, without requiring any changes to the user code.

---
<!-- Note: section break (`---`) onwards is not in CI merge commit -->

## Notes for Reviewers
* besides being a prerequisite for full ORC support, this change also simplifies #550

<!--
Pull Request(PR) Help

Before Merge Ensure:
* title reads like a short changelog line entry
* code includes tests and is documented
* leave the source better than before, but split out big reformats

See contributor (guide)[https://nim-works.github.io/nimskull/contributing.html]
for details, especially if you're new to this project.

Tips that make PRs easier:
* for big/impactful changes, start with chat/discussions to refine ideas
* refine the pull request message over time; don't have to nail it in one go
* handle the single commit message requirement at the end of review

saem commented on compiler/backend/jsgen.nim:
```diff
@@ -1041,6 +1044,14 @@ proc generateHeader(p: PProc, typ: PType): Rope =
 
+  if typ.callConv == ccClosure:
+    # generate the name for the hidden environment parameter. When creating a
+    # closure object, the environment is bound the ``this`` argument of the
```
```suggestion
    # closure object, the environment is bound to the ``this`` argument of the
```

zerbina replied:
Done, thanks.

saem reviewed (approved):
This is awesome, good to merge! 🚀 

I thought of two follow-ups. I'll start on the first one ~tomorrow and see about the second one based on the outcome:
1. have as much of `tests/lang_callable/iter` run for js/all platforms see where things land
2. quick update of the manual to remove the "big" mentions of can't use closure iterators for JS (rest can be long tail)

zerbina committed:
enable lambda-lifting for the JavaScript target

Summary
=======

Previously, inner closure procedures were not lifted into free-standing
procedures during `transf`, but rather kept as is and mapped to nested
JavaScript functions by `jsgen`. Besides requiring the JavaScript target
to be special-cased in `transf` and `lambdalifting`, this also caused
observable semantic issues: the destructors for captured variables were
called when leaving the scope in which they were defined, not when the
closure got destroyed.

In order to bring the closure semantics in line with the other targets
and as a prerequisite for full ORC support, lambda-lifting is now also
enabled when using the JS target -- the code generator is adjusted to
support the lifted procedures. The code generator changes are also a
preparation for full closure iterator support.

As a side-effect, closures now also work as expected when using the JS
target and executing non-compile-time-only procedures that have inner
closure procedures at compile-time.

Details
=======

- remove the dispatching based on the selected backend from `transf` and
  `lambdalifting`
- remove all code related to the previous closure support from `jsgen`
- implement support for lifted procedures in `jsgen`

The hidden environment parameter of routines using the `.closure`
calling convention is mapped to the this parameter at the JavaScript
level.

When a NimSkull closure object is created, a new function instance is
created via `bind` and the provided environment bound to the `this`
parameter. In addition, the both the function and environment are also
assigned to the respective prc and env properties on the function
object -- this allows for later retrieval.

Compared to using an object, this approach has the benefit that closures
can still be passed to an imported procedure that expects a JavaScript
function object, without requiring any changes to the user code.

zerbina commented:
> 1. have as much of `tests/lang_callable/iter` run for js/all platforms see where things land

Yep, that's a good idea. Having the `iter` tests work, at least in theory, on the JS and VM would be a big help when implementing closure iterator in the respective backends.

zerbina commented:
bors r+
